dss/operations/__init__.py

```diff
@@ -0,0 +1,103 @@
+"""
+This file provides an interface to build the DSS operations CLI, parts of which may be
+executed "in the cloud", via Lambda-SQS integration.
+
+Commands are defined with a target-action model. While command arguments configured at the
+target level are moved into the action arguments.
+
+e.g
+```
+dispatch = DSSOperationsCommandDispatch()
+
+storage = dispatch.target("storage", arguments={"replica", dict(choices=["dev", "integration", "staging", "prod"])}
+
+@storage.action("verify-referential-integrity")
+def verify_ref_integrity(argv, args):
+    ...
+
+# execution:
+scripts/dss-ops.py storage verify-referential-integrity --replica staging
+```
+"""
+import os
+import argparse
+import logging
+import traceback
+from functools import partial
+from uuid import uuid4
+
+import dss
+from dss import Replica
+
+logger = logging.getLogger(__name__)
+if not os.environ.get("DSS_VERSION"):  # detect non-lambda environment
+    logger.setLevel(logging.DEBUG)
+    ch = logging.StreamHandler()
+    ch.setLevel(logging.DEBUG)
+    logger.addHandler(ch)
+
+class _target:
+    def __init__(self, target_name, dispatcher):
+        self.target_name = target_name
+        self.dispatch = dispatch
+
+    def action(self, name: str, *, arguments: dict=None, mutually_exclusive: list=None):
+        dispatch = self.dispatch
+        arguments = arguments or dict()
+        if mutually_exclusive is None:
+            mutually_exclusive = dispatch.targets[self.target_name]['mutually_exclusive'] or list()
+
+        def register_action(obj):
+            parser = dispatch.targets[self.target_name]['subparser'].add_parser(name, help=obj.__doc__)
+            action_arguments = dispatch.targets[self.target_name]['arguments'].copy()
+            action_arguments.update(arguments)
+            for argname, kwargs in action_arguments.items():
+                if argname not in mutually_exclusive:
+                    parser.add_argument(argname, **(kwargs or dict()))
+            if mutually_exclusive:
+                group = parser.add_mutually_exclusive_group(required=True)
+                for argname in mutually_exclusive:
+                    kwargs = action_arguments.get(argname) or dict()
+                    group.add_argument(argname, **kwargs)
+            parser.add_argument("--job-id", default=dispatch.job_id)
+            parser.set_defaults(func=obj)
+            dispatch.actions[obj] = dict(target=dispatch.targets[self.target_name], name=name)
+            return obj
+        return register_action
+
+class DSSOperationsCommandDispatch:
+    """
+    Central dispatch for the DSS Operations CLI.
+    """
+
+    targets: dict = dict()
+    actions: dict = dict()
+
+    def __init__(self):
+        self.parser = argparse.ArgumentParser(description=self.__doc__)
+        self.parser_targets = self.parser.add_subparsers()
+        self.job_id = str(uuid4())
```

Comment:
How are job ids used?

Reply:
It's used as a correlation id in log messages. Makes it easy to find results in cloudwatch